File: apps/backend-java/src/main/java/com/housingplatform/features/properties/PropertySearchService.java
package com.housingplatform.features.properties;

import com.housingplatform.shared.auth.error.BadRequestException;
import com.housingplatform.persistence.repository.PropertyRepository;
import com.housingplatform.persistence.repository.PropertyRepositoryCustom.SearchPropertyRow;
import com.housingplatform.features.properties.dto.PropertySearchQuery;
import com.housingplatform.features.properties.dto.PropertySearchResult;
import com.housingplatform.features.properties.dto.PropertySearchSort;
import com.housingplatform.features.properties.mapper.PropertyMapper;
import com.housingplatform.features.properties.model.PropertySearchCriteria;
import com.housingplatform.shared.StorageUrlResolver;
import java.util.List;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class PropertySearchService {

  private final PropertyRepository propertyRepository;
  private final StorageUrlResolver storageUrlResolver;

  public PropertySearchService(
      PropertyRepository propertyRepository, StorageUrlResolver storageUrlResolver) {
    this.propertyRepository = propertyRepository;
    this.storageUrlResolver = storageUrlResolver;
  }

  @Transactional(readOnly = true)
  public PropertySearchResult search(PropertySearchQuery query) {
    validateSearchQuery(query);
    PropertySearchCriteria criteria = toCriteria(query);
    List<SearchPropertyRow> rows =
        propertyRepository.search(criteria, query.limit(), query.offset());
    int totalCount = rows.isEmpty() ? 0 : rows.getFirst().totalCount();
    var items = rows.stream().map(row -> PropertyMapper.toSearchCard(row, storageUrlResolver)).toList();
    return new PropertySearchResult(items, totalCount);
  }

  static PropertySearchCriteria toCriteria(PropertySearchQuery query) {
    Integer stayNights = null;
    if (query.checkIn() != null && query.checkOut() != null) {
      stayNights = (int) query.checkIn().datesUntil(query.checkOut()).count();
    }

    boolean hasBounds =
        query.north() != null
            && query.south() != null
            && query.east() != null
            && query.west() != null;

    String textQuery =
        query.query() == null || query.query().isBlank() ? null : query.query().strip();

    PropertySearchSort sort = query.sort() == null ? PropertySearchSort.recommended : query.sort();

    return new PropertySearchCriteria(
        textQuery,
        query.propertyType(),
        query.priceMin(),
        query.priceMax(),
        query.guests(),
        stayNights,
        query.checkIn(),
        sort,
        query.centerLat(),
        query.centerLng(),
        query.north(),
        query.south(),
        query.east(),
        query.west(),
        hasBounds,
        query.maxStationWalkMin(),
        query.amenitySlugs() == null || query.amenitySlugs().isEmpty()
            ? null
            : query.amenitySlugs(),
        query.excludePropertyIds() == null || query.excludePropertyIds().isEmpty()
            ? null
            : query.excludePropertyIds());
  }

  private static void validateSearchQuery(PropertySearchQuery query) {
    if (query.priceMin() != null
        && query.priceMax() != null
        && query.priceMax() < query.priceMin()) {
      throw new BadRequestException("priceMax must be greater than or equal to priceMin");
    }
    if (query.checkIn() != null
        && query.checkOut() != null
        && !query.checkOut().isAfter(query.checkIn())) {
      throw new BadRequestException("checkOut must be after checkIn");
    }
    if (query.maxStationWalkMin() != null && query.maxStationWalkMin() <= 0) {
      throw new BadRequestException("maxStationWalkMin must be greater than 0");
    }
  }
}

File: apps/backend-node/src/features/properties/property-search.service.ts
import { Injectable } from '@nestjs/common';

import { BadRequestError } from '../../shared/errors';
import { StorageUrlResolver } from '../../shared/storage/storage-url.resolver';
import type { PropertySearchQuery } from './dto/property-search-query';
import type { PropertySearchResult } from './dto/property-search-result';
import { PropertySearchSort } from './dto/property-search-sort';
import { mapSearchPropertyCard } from './mappers/property.mapper';
import { PropertyRepository } from './property.repository';
import { toSearchCriteria } from './property-search.criteria';

@Injectable()
export class PropertySearchService {
  constructor(
    private readonly repository: PropertyRepository,
    private readonly storage: StorageUrlResolver,
  ) {}

  async search(query: PropertySearchQuery): Promise<PropertySearchResult> {
    this.validateSearchQuery(query);

    const limit = query.limit ?? 20;
    const offset = query.offset ?? 0;
    const criteria = toSearchCriteria({
      ...query,
      sort: query.sort ?? PropertySearchSort.Recommended,
      limit,
      offset,
    });

    const rows = await this.repository.search(criteria, limit, offset);
    const totalCount = rows[0]?.totalCount ?? 0;
    const items = rows.map((row) => mapSearchPropertyCard(row, this.storage));

    return { items, totalCount };
  }

  private validateSearchQuery(query: PropertySearchQuery): void {
    if (
      query.priceMin != null &&
      query.priceMax != null &&
      query.priceMax < query.priceMin
    ) {
      throw new BadRequestError(
        'priceMax must be greater than or equal to priceMin',
      );
    }

    if (query.checkIn && query.checkOut && query.checkOut <= query.checkIn) {
      throw new BadRequestError('checkOut must be after checkIn');
    }

    if (query.maxStationWalkMin != null && query.maxStationWalkMin <= 0) {
      throw new BadRequestError('maxStationWalkMin must be greater than 0');
    }
  }
}

export function parsePropertySearchQuery(
  raw: Record<string, string | string[] | undefined>,
): PropertySearchQuery {
  const first = (value: string | string[] | undefined): string | undefined => {
    if (Array.isArray(value)) {
      return value[0];
    }
    return value;
  };

  const list = (value: string | string[] | undefined): string[] | undefined => {
    if (value == null) {
      return undefined;
    }
    if (Array.isArray(value)) {
      return value.flatMap((item) => item.split(',')).filter(Boolean);
    }
    return value.split(',').filter(Boolean);
  };

  const intOrNull = (value: string | undefined): number | null => {
    if (value == null || value.trim() === '') {
      return null;
    }
    const parsed = Number.parseInt(value, 10);
    return Number.isNaN(parsed) ? null : parsed;
  };

  const floatOrNull = (value: string | undefined): number | null => {
    if (value == null || value.trim() === '') {
      return null;
    }
    const parsed = Number.parseFloat(value);
    return Number.isNaN(parsed) ? null : parsed;
  };

  const sortValue = first(raw.sort);
  const sort =
    sortValue &&
    Object.values(PropertySearchSort).includes(sortValue as PropertySearchSort)
      ? (sortValue as PropertySearchSort)
      : PropertySearchSort.Recommended;

  return {
    query: first(raw.query) ?? null,
    propertyType: first(raw.propertyType) as PropertySearchQuery['propertyType'],
    checkIn: first(raw.checkIn) ?? null,
    checkOut: first(raw.checkOut) ?? null,
    guests: intOrNull(first(raw.guests)),
    priceMin: intOrNull(first(raw.priceMin)),
    priceMax: intOrNull(first(raw.priceMax)),
    sort,
    centerLat: floatOrNull(first(raw.centerLat)),
    centerLng: floatOrNull(first(raw.centerLng)),
    north: floatOrNull(first(raw.north)),
    south: floatOrNull(first(raw.south)),
    east: floatOrNull(first(raw.east)),
    west: floatOrNull(first(raw.west)),
    amenitySlugs: list(raw.amenitySlugs) ?? null,
    maxStationWalkMin: intOrNull(first(raw.maxStationWalkMin)),
    excludePropertyIds: list(raw.excludePropertyIds) ?? null,
    limit: intOrNull(first(raw.limit)) ?? 20,
    offset: intOrNull(first(raw.offset)) ?? 0,
  };
}
